[Flaky-test] MailboxServiceTest.testRemoteEarlyTerminated

```
[INFO] Running org.apache.pinot.query.mailbox.MailboxServiceTest
12:42:02.146 WARN [MailboxContentObserver] [grpc-default-executor-2] Timed out adding block into mailbox: 8|1|0|0|0 with timeout: 656ms
12:42:02.152 WARN [MailboxContentObserver] [grpc-default-executor-2] Error on receiver side
io.grpc.StatusRuntimeException: CANCELLED: client cancelled
	at io.grpc.Status.asRuntimeException(Status.java:524) ~[grpc-api-1.72.0.jar:1.72.0]
	at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:291) [grpc-stub-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closedInternal(ServerCallImpl.java:375) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closed(ServerCallImpl.java:364) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1Closed.runInContext(ServerImpl.java:910) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) [grpc-core-1.72.0.jar:1.72.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
12:42:02.140 WARN [MailboxStatusObserver] [grpc-default-executor-1] Error on sender side
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: CallOptions deadline exceeded after 0.885505252s. Name resolution delay 0.008412377 seconds. [closed=[], committed=[buffered_nanos=137753819, remote_addr=localhost/127.0.0.1:45531]]
	at io.grpc.Status.asRuntimeException(Status.java:532) ~[grpc-api-1.72.0.jar:1.72.0]
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:564) [grpc-stub-1.72.0.jar:1.72.0]
	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:489) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:453) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:486) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:565) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl.access$100(ClientCallImpl.java:72) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:733) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:714) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) [grpc-core-1.72.0.jar:1.72.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
12:42:03.323 WARN [MailboxStatusObserver] [grpc-default-executor-2] Error on sender side
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: CallOptions deadline exceeded after 0.998949235s. Name resolution delay 0.000000000 seconds. [closed=[], committed=[remote_addr=localhost/127.0.0.1:45531]]
	at io.grpc.Status.asRuntimeException(Status.java:532) ~[grpc-api-1.72.0.jar:1.72.0]
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:564) [grpc-stub-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:565) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl.access$100(ClientCallImpl.java:72) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:733) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:714) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) [grpc-core-1.72.0.jar:1.72.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
12:42:03.324 WARN [MailboxContentObserver] [grpc-default-executor-1] Error on receiver side
io.grpc.StatusRuntimeException: CANCELLED: client cancelled
	at io.grpc.Status.asRuntimeException(Status.java:524) ~[grpc-api-1.72.0.jar:1.72.0]
	at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:291) [grpc-stub-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closedInternal(ServerCallImpl.java:375) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closed(ServerCallImpl.java:364) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1Closed.runInContext(ServerImpl.java:910) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [grpc-core-1.72.0.jar:1.72.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) [grpc-core-1.72.0.jar:1.72.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
Error:  Tests run: 16, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 15.63 s <<< FAILURE! -- in org.apache.pinot.query.mailbox.MailboxServiceTest
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalBufferFull -- Time elapsed: 1.011 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalCancelledByReceiver -- Time elapsed: 0 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalCancelledBySender -- Time elapsed: 0.001 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalCancelledBySenderBeforeSend -- Time elapsed: 0 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalEarlyTerminated -- Time elapsed: 0.001 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalHappyPathReceiveFirst -- Time elapsed: 0 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalHappyPathSendFirst -- Time elapsed: 0.001 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testLocalTimeOut -- Time elapsed: 1.012 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteBufferFull -- Time elapsed: 1.034 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteCancelledByReceiver -- Time elapsed: 0.006 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteCancelledBySender -- Time elapsed: 0.010 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteCancelledBySenderBeforeSend -- Time elapsed: 0.005 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteEarlyTerminated -- Time elapsed: 0.010 s <<< FAILURE!
java.lang.AssertionError: expected object to not be null
	at org.testng.Assert.fail(Assert.java:111)
	at org.testng.Assert.assertNotNull(Assert.java:1496)
	at org.testng.Assert.assertNotNull(Assert.java:1480)
	at org.apache.pinot.query.mailbox.MailboxServiceTest.readBlock(MailboxServiceTest.java:632)
	at org.apache.pinot.query.mailbox.MailboxServiceTest.lambda$testRemoteEarlyTerminated$11(MailboxServiceTest.java:607)
	at org.apache.pinot.util.TestUtils.waitForCondition(TestUtils.java:104)
	at org.apache.pinot.util.TestUtils.waitForCondition(TestUtils.java:90)
	at org.apache.pinot.util.TestUtils.waitForCondition(TestUtils.java:95)
	at org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteEarlyTerminated(MailboxServiceTest.java:606)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:141)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:687)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:230)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:63)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:995)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:203)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:154)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:134)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.testng.TestRunner.privateRun(TestRunner.java:741)
	at org.testng.TestRunner.run(TestRunner.java:616)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:421)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:413)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:373)
	at org.testng.SuiteRunner.run(SuiteRunner.java:312)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1274)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1208)
	at org.testng.TestNG.runSuites(TestNG.java:1112)
	at org.testng.TestNG.run(TestNG.java:1079)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:155)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:102)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:91)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:137)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteHappyPathReceiveFirst -- Time elapsed: 0.029 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteHappyPathSendFirst -- Time elapsed: 0.104 s
Error:  org.apache.pinot.query.mailbox.MailboxServiceTest.testRemoteTimeOut -- Time elapsed: 1.010 s
```

--- a/pinot-query-runtime/src/test/java/org/apache/pinot/query/mailbox/MailboxServiceTest.java
+++ b/pinot-query-runtime/src/test/java/org/apache/pinot/query/mailbox/MailboxServiceTest.java
@@ -22,6 +22,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
 import org.apache.pinot.common.datatable.StatMap;
 import org.apache.pinot.common.utils.DataSchema;
 import org.apache.pinot.common.utils.DataSchema.ColumnDataType;
@@ -604,7 +605,11 @@ public class MailboxServiceTest {
     sendingMailbox.send(OperatorTestUtil.block(DATA_SCHEMA, new Object[]{0}));
     // receiving-side early terminates after pulling the first block
     TestUtils.waitForCondition(aVoid -> {
-      MseBlock block = readBlock(receivingMailbox);
+      ReceivingMailbox.MseBlockWithStats blockWithStats = receivingMailbox.poll();
+      if (blockWithStats == null) {
+        return false;
+      }
+      MseBlock block = blockWithStats.getBlock();
       return block != null && block.isData() && ((MseBlock.Data) block).getNumRows() == 1;
     }, 1000L, "Failed to deliver mails");
     receivingMailbox.earlyTerminate();
@@ -627,9 +632,12 @@ public class MailboxServiceTest {
     return rows;
   }
 
+  @Nullable
   public static MseBlock readBlock(ReceivingMailbox receivingMailbox) {
     ReceivingMailbox.MseBlockWithStats block = receivingMailbox.poll();
-    assertNotNull(block);
+    if (block == null) {
+      return null;
+    }
     return block.getBlock();
   }
 }

